### ISSUE TITLE
_generation token mask error

### ISSUE THREAD

**Issue by xueyan-lii** [pytorch/torchtune#2590]
In generation/_generation.py, in the last step, 
```
if stop_tokens is not None:
        generated_tokens *= stop_token_mask
        generated_logits *= stop_token_mask[:, -generated_logits.shape[1] :, None]
```
there is an error where generated_tokens are converted to 0 by the stop_token_mask at the positions that have already hit the stop token. However, this is wrong because token 0 will be decoded to a symbol like '!' rather than ignored in the output. This is why running the evaluation script with batch size larger than 1 will result in all but one generation with trailing symbols. However, this generally does not affect the evalution performance as lm_eval will extract the correct answer positions most of the time.

The correct way should be to convert trailing tokens to the pad_id, so they will be ignored during decoding.
```
if stop_tokens is not None:
        # Create a mask where True indicates positions after a stop token
        mask_to_pad = ~stop_token_mask.bool()
        # Replace these positions with the padding token
        generated_tokens = torch.where(
            mask_to_pad, 
            torch.tensor(pad_id, device=generated_tokens.device, dtype=generated_tokens.dtype), 
            generated_tokens
        )
        generated_logits *= stop_token_mask[:, -generated_logits.shape[1] :, None]
```

**Comment by joecummings** [comment]
Good catch! We'd love to accept a PR if you put once up to fix this :)

Please feel free to tag me for review.
